presto-native-execution/presto_cpp/main/operators/tests/UnsaferowShuffleTest.cpp

```diff
@@ -0,0 +1,448 @@
+    // hold row size.
+    BufferPtr buffer = AlignedBuffer::allocate<char>(
+        totalSize + sizeof(size_t) * serializedData->size(), pool());
+    auto rawBuffer = buffer->asMutable<char>();
+
+    // Copy data.
+    size_t offset = 0;
+    for (auto i = 0; i < serializedData->size(); ++i) {
+      auto value = serializedData->valueAt(i);
+
+      *(size_t*)(rawBuffer + offset) = value.size();
+      offset += sizeof(size_t);
+
+      memcpy(rawBuffer + offset, value.data(), value.size());
+      offset += value.size();
+    }
+
+    // Deserialize the block.
+    return deserialize(buffer, rowType);
+  }
+
+  RowVectorPtr deserialize(BufferPtr& serialized, const RowTypePtr& rowType) {
+    auto serializer =
+        std::make_unique<serializer::spark::UnsafeRowVectorSerde>();
+
+    ByteRange byteRange = {
+        serialized->asMutable<uint8_t>(), (int32_t)serialized->size(), 0};
+
+    auto input = std::make_unique<ByteStream>();
+    input->resetInput({byteRange});
+
+    RowVectorPtr result;
+    serializer->deserialize(input.get(), pool(), rowType, &result, nullptr);
+    return result;
+  }
+
+  std::shared_ptr<const RowType> rowType(
+      std::vector<std::shared_ptr<const Type>>&& types) {
+    std::vector<std::string> names;
+    for (int32_t i = 0; i < types.size(); ++i) {
+      names.push_back(fmt::format("c{}", i));
+    }
+    return ROW(std::move(names), std::move(types));
+  }
+
+  RowVectorPtr rowVector(const std::vector<VectorPtr>& children) {
+    std::vector<std::string> names;
+    for (int32_t i = 0; i < children.size(); ++i) {
+      names.push_back(fmt::format("c{}", i));
+    }
+
+    return rowVector(std::move(names), children);
+  }
+
+  RowVectorPtr rowVector(
+      std::vector<std::string> childNames,
+      const std::vector<VectorPtr>& children) {
+    std::vector<std::shared_ptr<const Type>> childTypes;
+    childTypes.resize(children.size());
+    for (int i = 0; i < children.size(); i++) {
+      childTypes[i] = children[i]->type();
+    }
+    auto rowType = ROW(std::move(childNames), std::move(childTypes));
+    const size_t vectorSize = children.empty() ? 0 : children.front()->size();
+
+    return std::make_shared<RowVector>(
+        pool_.get(), rowType, BufferPtr(nullptr), vectorSize, children);
+  }
+
+  RowVectorPtr rowVector(
+      const std::shared_ptr<const RowType>& rowType,
+      vector_size_t size) {
+    return std::make_shared<RowVector>(
+        pool_.get(),
+        rowType,
+        BufferPtr(nullptr),
+        size,
+        std::vector<VectorPtr>{});
+  }
+
+  RowVectorPtr makeRowVector(
+      const std::vector<VectorPtr>& children,
+      std::function<bool(vector_size_t /*row*/)> isNullAt = nullptr) {
+    auto rowVec = rowVector(children);
+    if (isNullAt) {
+      setNulls(rowVec, isNullAt);
+    }
+    return rowVec;
+  }
+
+  void assertEqualVectors(const VectorPtr& expected, const VectorPtr& actual) {
```

Comment:
Can we use the existing function from velox/vector/tests/VectorTestBase.h ?

Reply:
I tried but given that it doesn't seem to be reference-able in presto_cpp

Reply:
> doesn't seem to be reference-able in presto_cpp

Hmm... what does that mean? What happens if you try using this function?

Reply:
The body of ```assertEqualVectors```  is located under ```VeloxVectorTest.cpp``` which is under ```velox_vector_test_lib```. When I add that lib to CMake file I get a link error:
```
ld: library not found for -lvelox_vector_test_lib
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Reply:
Aha... This needs to be fixed.

In Velox, there are 2 build variables that control what test code is built: VELOX_BUILD_TESTING and VELOX_BUILD_TEST_UTILS. VELOX_BUILD_TESTING is intended to allow for disabling building of tests (to speed up builds). VELOX_BUILD_TEST_UTILS is intended to allow for disabling building test utilities. 

VectorTestBase.* files are test utilities, however, their build is guarded by VELOX_BUILD_TESTING, not VELOX_BUILD_TEST_UTILS.

presto-native-execution/velox/velox/vector/CMakeLists.txt

```
if(${VELOX_BUILD_TESTING} OR ${VELOX_ENABLE_BENCHMARKS_BASIC})
  add_subdirectory(tests)
  add_subdirectory(benchmarks)
endif()
```

This is incorrect as we should check VELOX_BUILD_TEST_UTILS variable when deciding whether to build VectorTestBase.* files (velox_vector_test_lib).

This is not a problem with velox_exec_test_util that contains OperatorTestBase.* because that is being build even if VELOX_BUILD_TESTING is off, but VELOX_BUILD_TEST_UTILS is on.

presto-native-execution/velox/velox/exec/CMakeLists.txt

```
if(${VELOX_BUILD_TESTING})
  add_subdirectory(tests)
  add_subdirectory(benchmarks)
elseif(${VELOX_BUILD_TEST_UTILS})
  add_subdirectory(tests/utils)
endif()
```

When building Prestissimo, we set VELOX_BUILD_TESTING OFF and VELOX_BUILD_TEST_UTILS ON.

presto-native-execution/CMakeLists.txt

```
set(VELOX_BUILD_TESTING
    OFF
    CACHE BOOL "Enable Velox tests")

option(PRESTO_ENABLE_TESTING "Enable tests" ON)

set(VELOX_BUILD_TEST_UTILS
    ${PRESTO_ENABLE_TESTING}
    CACHE BOOL "Enable Velox test utils")
```

A fix could be to move to call `add_subdirectory(tests)` unconditionally in vector/CMakeLists.txt, then build velox_vector_test_lib if VELOX_BUILD_TEST_UTILS is ON and the rest of the tests if VELOX_BUILD_TESTING is ON. Alternatively, we could move VectorMaker.* and VectorTestBase.* into a separate directory and call `add_directory` for that if VELOX_BUILD_TEST_UTILS is ON.

CC: @majetideepak @kgpai

Reply:
@mbasmanova I tried the solution #1 and it worked! So I'm using vector maker and also assertEqualityVectors ... in this PR as well. Preparing a Velox PR for that change too.

Reply:
This method is not needed. Please, remove.